spec/outputs/s3_spec.rb

```diff
@@ -1,6 +1,322 @@
+      Stud::Temporary.file("logstash", "a+") do |tmp|
+        s3 = LogStash::Outputs::S3.new(minimal_settings)
+        allow(s3).to receive(:test_s3_write)
+        s3.register
+        s3.tempfile = tmp
+        s3.write_to_tempfile("test-write")
+        tmp.rewind
+        expect(tmp.read).to eq("test-write")
+      end
+    end
+  end
+
+  describe "#rotate_events_log" do
+    let(:s3) { LogStash::Outputs::S3.new(minimal_settings.merge({ "size_file" => 1024 })) }
+
+    it "returns true if the tempfile is over the file_size limit" do
+      Stud::Temporary.file do |tmp|
+        tmp.stub(:size) { 2024001 }
+
+        s3.tempfile = tmp
+        expect(s3.rotate_events_log?).to be(true)
+      end
+    end
+
+    it "returns false if the tempfile is under the file_size limit" do
+      Stud::Temporary.file do |tmp|
+        tmp.stub(:size) { 100 }
+
+        s3.tempfile = tmp
+        expect(s3.rotate_events_log?).to eq(false)
+      end
+    end
+  end
+
+  describe "#move_file_to_bucket" do
+    let!(:s3) { LogStash::Outputs::S3.new(minimal_settings) }
+
+    before do
+      # Assume the AWS test credentials pass.
+      allow(s3).to receive(:test_s3_write)
+      s3.register
+    end
+
+    it "should always delete the source file" do
+      tmp = Stud::Temporary.file
+
+      allow(File).to receive(:zero?).and_return(true)
+      expect(File).to receive(:delete).with(tmp)
+
+      s3.move_file_to_bucket(tmp)
+    end
+
+    it 'should not upload the file if the size of the file is zero' do
+      temp_file = Stud::Temporary.file
+      allow(temp_file).to receive(:zero?).and_return(true)
+
+      expect(s3).not_to receive(:write_on_bucket)
+      s3.move_file_to_bucket(temp_file)
+    end
+
+    it "should upload the file if the size > 0" do
+      tmp = Stud::Temporary.file
+
+      allow(File).to receive(:zero?).and_return(false)
+      expect(s3).to receive(:write_on_bucket)
+
+      s3.move_file_to_bucket(tmp)
+    end
+  end
+
+  describe "#restore_from_crashes" do
+    it "read the temp directory and upload the matching file to s3" do
+      s3 = LogStash::Outputs::S3.new(minimal_settings.merge({ "temporary_directory" => "/tmp/logstash/" }))
+
+      expect(Dir).to receive(:[]).with("/tmp/logstash/*.txt").and_return(["/tmp/logstash/01.txt"])
+      expect(s3).to receive(:move_file_to_bucket_async).with("/tmp/logstash/01.txt")
+
+
+      s3.restore_from_crashes
+    end
+  end
+
+  describe "#receive" do
+    it "should send the event through the codecs" do
+      data = {"foo" => "bar", "baz" => {"bah" => ["a","b","c"]}, "@timestamp" => "2014-05-30T02:52:17.929Z"}
+      event = LogStash::Event.new(data)
+
+      expect_any_instance_of(LogStash::Codecs::Line).to receive(:encode).with(event)
+
+      s3 = LogStash::Outputs::S3.new(minimal_settings)
+      allow(s3).to receive(:test_s3_write)
+      s3.register
+
+      s3.receive(event)
+    end
+  end
+
+  describe "when rotating the temporary file" do
```

Comment:
It requires a local connection for me, you should probably mock the connection parts.